source/compiler-core/slang-json-parser.cpp

```diff
@@ -0,0 +1,384 @@
+// slang-json-parser.cpp
+#include "slang-json-parser.h"
+
+#include "slang-json-diagnostics.h"
+
+/*
+https://www.json.org/json-en.html
+*/
+
+namespace Slang {
+
+SlangResult JSONParser::_parseObject()
+{
+    SLANG_RETURN_ON_FAIL(m_lexer->expect(JSONTokenType::LBrace));
+
+    m_listener->startObject();
+
+    if (m_lexer->advanceIf(JSONTokenType::RBrace))
+    {
+        m_listener->endObject();
+        return SLANG_OK;
+    }
+
+    while (true)
+    {
+        JSONToken keyToken;
+        SLANG_RETURN_ON_FAIL(m_lexer->expect(JSONTokenType::StringLiteral, keyToken));
+        m_listener->addLexemeKey(m_lexer->getLexeme(keyToken));
+
+        SLANG_RETURN_ON_FAIL(m_lexer->expect(JSONTokenType::Colon));
+
+        SLANG_RETURN_ON_FAIL(_parseValue());
+        if (m_lexer->advanceIf(JSONTokenType::Comma))
+        {
+            continue;
+        }
+
+        break;
+    }
+
+    SLANG_RETURN_ON_FAIL(m_lexer->expect(JSONTokenType::RBrace));
+    m_listener->endObject();
+    return SLANG_OK;
+}
+
+SlangResult JSONParser::_parseArray()
+{
+    SLANG_RETURN_ON_FAIL(m_lexer->expect(JSONTokenType::LBracket));
+
+    m_listener->startArray();
+
+    if (m_lexer->advanceIf(JSONTokenType::RBracket))
+    {
+        m_listener->endArray();
+        return SLANG_OK;
+    }
+
+    while (true)
+    {
+        SLANG_RETURN_ON_FAIL(_parseValue());
+        if (m_lexer->advanceIf(JSONTokenType::Comma))
+        {
+            continue;
+        }
+        break;
+    }
+
+    SLANG_RETURN_ON_FAIL(m_lexer->expect(JSONTokenType::RBracket));
+    m_listener->endArray();
+    return SLANG_OK;
+}
+
+SlangResult JSONParser::_parseValue()
+{
+    switch (m_lexer->peekType())
+    {
+        case JSONTokenType::True:
+        case JSONTokenType::False:
+        case JSONTokenType::Null:
+        case JSONTokenType::IntegerLiteral:
+        case JSONTokenType::FloatLiteral:
+        case JSONTokenType::StringLiteral:
+        {
+            m_listener->addLexemeValue(m_lexer->peekType(), m_lexer->peekLexeme());
+            m_lexer->advance();
+            return SLANG_OK;
+        }
+        case JSONTokenType::LBracket:
+        {
+            return _parseArray();
+        }
+        case JSONTokenType::LBrace:
+        {
+            return _parseObject();
+        }
+        default:
+        {
+            m_sink->diagnose(m_lexer->peekLoc(), JSONDiagnostics::unexpectedToken, getJSONTokenAsText(m_lexer->peekType()));
+            return SLANG_FAIL;
+        }
+        case JSONTokenType::Invalid:
+        {
+            // It's a lex error, so just fail
+            return SLANG_FAIL;
+        }
+    }
+}
+
+SlangResult JSONParser::parse(JSONLexer* lexer, SourceView* sourceView, JSONListener* listener, DiagnosticSink* sink)
```

Comment:
A small design question is whether the decomposition into `JSONLexer` and `JSONParser` pays off, or if we could write a single combined lexer/parser (a so-called "scannerless" parser) that handles both tasks. This is akin to how many Lisp implementations minimize code size/complexity by combining lexing and parsing.